File: frontend/src/lib/schedule.ts
import { fromISO } from "./dates";

export const WEEKDAYS = ["Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Нд"];
export const EVERY_DAY = [0, 1, 2, 3, 4, 5, 6];

export interface Schedule {
  start_date?: string | null;
  weekdays?: readonly number[];
  archived_at?: string | null;
}

/** Розклад враховує дати, коли звичка вже існувала і ще не була архівована. */
export function isScheduledOn(habit: Schedule, day: string): boolean {
  if (habit.start_date && day < habit.start_date) return false;
  if (habit.archived_at && day > habit.archived_at) return false;
  return (habit.weekdays ?? EVERY_DAY).includes((fromISO(day).getDay() + 6) % 7);
}

export function scheduleLabel(habit: Schedule): string {
  const days = habit.weekdays ?? EVERY_DAY;
  return days.length === 7 ? "Щодня" : days.map((day) => WEEKDAYS[day]).join(", ");
}

export function pluralCompletions(count: number): string {
  const last = count % 10;
  const lastTwo = count % 100;
  if (lastTwo >= 11 && lastTwo <= 14) return "запланованих виконань";
  if (last === 1) return "заплановане виконання";
  if (last >= 2 && last <= 4) return "заплановані виконання";
  return "запланованих виконань";
}

File: bot/schedule.py
"""Розклад звички з відповіді HTTP API, без доступу бота до бази."""

from datetime import date


def is_planned_on(habit: dict, day: date) -> bool:
    """Чи запланована активна звичка на день у часовому поясі власника."""
    if habit.get("archived_at") is not None:
        return False
    start = habit.get("start_date")
    if start and day < date.fromisoformat(start):
        return False
    return day.weekday() in habit.get("weekdays", range(7))


WEEKDAY_NAMES = ("Пн", "Вт", "Ср", "Чт", "Пт", "Сб", "Нд")
EVERY_DAY = list(range(7))
WORKDAYS = list(range(5))


def schedule_label(weekdays: list[int]) -> str:
    """«Щодня» або «Пн, Ср, Пт» — той самий підпис, що й у вебі (scheduleLabel)."""
    days = sorted(set(weekdays))
    if len(days) == 7:
        return "Щодня"
    return ", ".join(WEEKDAY_NAMES[day] for day in days)


def toggle_day(weekdays: list[int], day: int) -> list[int]:
    """Увімкнути або вимкнути день. Відсортовано, бо API вимагає впорядкований список."""
    days = set(weekdays)
    days.symmetric_difference_update({day})
    return sorted(days)
